utils/utils.py

```diff
@@ -71,55 +72,188 @@ def train_validate_test_normal(
             f"Test RMSE: {test_rmse[0]:.8f}"
         )
+        print("Tasks MAE:", train_taskserr)
+
+        trainlib.append(train_mae)
+        vallib.append(val_mae)
+        testlib.append(test_rmse[0])
+        tasklib.append(train_taskserr)
+        tasklib_vali.append(val_taskserr)
+        tasklib_test.append(test_rmse[1])
+
+        tasklib_nodes.append(train_taskserr_nodes)
+        tasklib_vali_nodes.append(val_taskserr_nodes)
+        tasklib_test_nodes.append(test_rmse[2])
+
+        ###tracking the solution evolving with training
+        if plot_hist_sol:
+            true_values = test_rmse[3]
+            predicted_values = test_rmse[4]
+            for ihead in range(model.num_heads):
+                visualizer = Visualizer(model_with_config_name)
+                visualizer.add_test_values(
+                    true_values=true_values[ihead],
+                    predicted_values=predicted_values[ihead],
+                )
+                if model.head_dims[ihead] // num_nodes == 3:  # magnetic moments
+                    visualizer.create_scatter_plot_atoms_vec(
+                        config["Target_dataset"]["output_names"][ihead],
+                        x_atomfeature,
+                        epoch,
+                    )
+                else:
+                    visualizer.create_scatter_plot_atoms(
+                        config["Target_dataset"]["output_names"][ihead],
+                        x_atomfeature,
+                        epoch,
+                    )
+                    visualizer.create_error_histogram_plot_atoms(
+                        config["Target_dataset"]["output_names"][ihead],
+                        x_atomfeature,
+                        epoch,
+                    )
+
     # At the end of training phase, do the one test run for visualizer to get latest predictions
-    visualizer = Visualizer(model_with_config_name)
-    test_rmse, true_values, predicted_values = test(
-        test_loader, model, config["output_dim"]
+    test_rmse, test_taskserr, test_taskserr_nodes, true_values, predicted_values = test(
+        test_loader, model, config["Architecture"]["output_dim"]
     )
+
     if (
-        config["denormalize_output"] == "True"
+        config["Target_dataset"]["denormalize_output"] == "True"
     ):  ##output predictions with unit/not normalized
-        y_minmax = config["y_minmax"]
-        for isamp in range(len(predicted_values)):
-            for iout in range(len(predicted_values[0])):
-                predicted_values[isamp][iout] = (
-                    predicted_values[isamp][iout]
-                    * (y_minmax[iout][1] - y_minmax[iout][0])
-                    + y_minmax[iout][0]
+        true_values, predicted_values = output_denormalize(
+            config["Target_dataset"]["y_minmax"], true_values, predicted_values
+        )
+
+    for ihead in range(model.num_heads):
```

Comment:
the plot of the solution should be a separate function. The visualization techniques should be part of the training.
In fact, I would be nice if the training function returned the history of training-valdiation-test-losses, and then these were used by another function